Weapon System : Gun shooting and Object Light Taking

fluetanol opened the pull request:
<!-- PR 제목은 간결하게 작성해주세요 -->

## ✨ 변경 사항
<!-- 변경 사항에 대한 설명을 적어주세요 -->

Weapon을 플레이어에게 적용시키는 과정은 내가 따로 하면 되기때문에 별다른 설명은 안하겠음
해당 코드 내용이 정 궁금하다면 **Script -> Taeseung->LaserGunBullet**폴더안의 클래스를 뜯어보거나,
**Prefab->Weapon_Prefab->Gun&Bullet**의 LaserBullet(탄알), PrototypeGun에 붙어 있는 
**LaserGunShootingManager**과 **LaserGunBulletManager** 스크립트 참고


### 현재 총이 되는 기능
1. 오브젝트로 부터 빛의 흡수
2. 총이 흡수한 빛의 양에 대한 시각적 표시
3. 총의 팀색깔에 따른 흡수 가능한 오브젝트 구분
4. 탄알 발사
5. 거울에 닿을시 탄알 반사
6. 탄알 발사 전, 경로상에 거울이 존재하면 반사될 경로를 시각적으로 미리 보여주는 기능

까지 가능, 다만 반사 메커니즘은 잘못하면 버그가 있을 것 같음
총은 아직까진 화면 아무데나 짧은 터치로 발사 가능함.
우진이 작업 머지하는 대로 버튼에 적용 할 예정

### 오브젝트 판정
현재까지 오브젝트는 두 부류로 나뉨
1. 거울
2. 흡수 가능한 빛 오브젝트

이 둘을 구분 짓는 기준은 레이어에 따라 달려있음
**거울 -> 6:Mirror
빛 오브젝트 -> 7:LightObject**
이 설정을 안해두면 밑에 설명할 스크립트를 아무리 붙혀도 아무런 효과 없으니 꼭 설정 해두십쇼


### 오브젝트 빛 흡수 관련 스크립트
빛을 흡수 받을 오브젝트 입장에선 Object Emission Manager를,
빛을 흡수 하는 플레이어 입장에선 Object Emission Take Manager를 붙혀야 함
**또한 빛을 뿜는 오브젝트는 머터리얼에 URP/Lit, 이미션 설정가능한 기능이(인자명이 _emission)이 반드시 존재해야함
정 애매하면 Materials -> Weapon_Material -> Light -> LaserLight 머터리얼 적용해서 사용**

해당 스크립트에 대한 자세한 내용은 Script TaeSeung폴더의 Object폴더 참고

이 스크립트를 테스트 하기 위해 기본 입력해야 할 인자 설명

**1. Object Emission Take Manager**
**Team :** 어떤 색의 팀인지 입력, 이를 기준으로 오브젝트가 흡수 가능한 색인지 판별할 예정
**LasergunManager :** 해당 스크립트가 붙어있는 무기 오브젝트를 끌어다 넣으면 됨. 이 스크립트는 무기로 쓸 오브젝트에 붙어있어야 하는데, 위에 설명한 무기 프리팹 경로를 참고해서 지금은 Prototypegun을 그냥 쓸 것을 권장
**Choiceoutlinematerial :** 이건 해당 오브젝트를 선택했을 때 그 오브젝트가 선택되었다는  외곽선 머터리얼을 넣으면 되는데, 아직까진 테스트 기능이니 무시해도 됨.

**2. Object Emission Manager**
**MeshRenderer:** 해당 스크립트가 붙은 오브젝트의 MeshRenderer을 넣으면 됨
**Colortype:** 이 오브젝트가 뿜을 컬러빛 설정
**Gauge:** 이 오브젝트가 가지고 있을 빛 게이지
**TakeWeight:** 이 오브젝트의 빛을 흡수할때 틱당 얼마나 흡수될지 설정하는 건데, 1이외의 값은 아직까지 버그 있고 불안정하게 작동하므로 그냥 1 넣자
**Emissionstrength** : 이 오브젝트가 뿜을 빛의 세기, 2~3정도가 적당해보임
**EmissionUI :** 이 오브젝트의 빛의 잔량을 표시해주는 UI스크립트, 일단은 **Prefab-> Weapon_Prefab->UI**에 들어있는 **Canvas**라는 이름의 오브젝트를 이 항목에 끌어다 넣으시면 됩니다. 그렇게 안하면 무조건 이상하게 작동함. 이 UI부분은 나중에 내가 동혁이 형이랑 같이 상의해서 제작하면 될듯.


또한, 빛 흡수를 위해선 해당 오브젝트를 1초가량 눌러야 빨아들입니다.


여기까지고,
사용예시를 정확히 알고 싶으면 SCENE 폴더의 Weapon_Sample Scene의 Laser Test Scene을 직접 사용해보시길 바랍니다.

잘 사용 될지 모르겠네
감사합니다.


## 📚 새로 알게 된 내용 혹은 궁금한 사항
<!-- 참고할 사항이 있다면 적어주세요 -->



## 📌 관련 이슈
<!-- 관련있는 이슈 번호(#000)을 적어주세요.
  해당 pull request merge와 함께 이슈를 닫으려면
  closed #Issue_number를 적어주세요 -->

fluetanol committed:
초기화 및 총 구현

1. 환경 설정
2. 총 발사 구현
3. 총 반사 구현
4. 총의 탄창으로 쓸 빛 에너지 흡수 구현
5. probuilder 추가
6. layer 6,7 mirror, lightobeject 추가 (레이어 안겹칙게 주의해주세요)

fluetanol committed:
Light Object UI making

발광하는 물체들에 대해 오브젝트가 갖고 있는 남은 불빛을 보여줌

fluetanol committed:
Temp commit(Object Emission UI Develop)

emission ui 개발 임시 커밋

fluetanol committed:
Object Emission UI making

오브젝트 클릭시 빛 잔량 표출해주는 ui생성

fluetanol committed:
Weapon Resource Folder Arrange

현재까지 한 무기관련 리소스들 통합 폴더에 정리 작업
무기 관련 리소스들은 전부 Weapon_ 으로 시작되는 폴더에 옮겨짐

fluetanol committed:
Merge branch 'develop' into TaeSeung

WoojinKim1225 commented:
일단 올리고 TakeWeight 1 아니면 버그 있는거 고칩시당

WoojinKim1225 reviewed (approved):
확인완료

taehyeoon commented:
## 질문사항
1. Taeseung 폴더 아래에 Script라는 폴더는 필요 없어보임(상위 폴더와 의미 중복)
2. Taeseung 폴더 아래에 shader 폴더는 Assets/Shader를 생성해서 옮기는 것이 좋아보임
3. Assets/Scenes/Weapon_SampleScene/SampleScene.unity 이 씬은 사용하는 씬인건지?
4. 애니메이션 폴더 내의 Weapon_PrototypeAnimation 폴더 명에서 뒤에 애니메이션은 삭제하는 것이 나아보임(상위 폴더와 중복되는 의미) + 메테리얼 폴더
5. PrototypeGun animationController에 애니메이션 이름이 모호함
6. 총 발사 애니메이션은 어디서 관리하는지?
7. GunHandlePrototype은 왜 2개인지? 다른 용도라면 이름 구분 필요
8. Mirror 프리팹에 포함된 카메라의 용도는 무엇인지?
9. GrassSound 프리팹은 사운드만 가지고 있는것 같은데 다른 방식으로 관리할 수는 없는지?
10. Canvas 프리팹의 이름이 모호함
11. Assets/Resources 폴더는 빌드때 반드시 포함되며, 코드에서도 접근할 수 있는 폴더인데, 반드시 텍스쳐가 이 폴더 안에 있어야 하는지? 아니라면 따로 Assets/Textures 폴더를 만드는 것은 어떤지?

taehyeoon reviewed (changes requested):
질문내용 참고 바랍니다

fluetanol commented:
Taeseung 폴더 아래에 Script라는 폴더는 필요 없어보임(상위 폴더와 의미 중복)
-> 해결

 Taeseung 폴더 아래에 shader 폴더는 Assets/Shader를 생성해서 옮기는 것이 좋아보임
-> 해결, 오히려 좋

Assets/Scenes/Weapon_SampleScene/SampleScene.unity 이 씬은 사용하는 씬인건지?
-> 더미였음 삭제함

애니메이션 폴더 내의 Weapon_PrototypeAnimation 폴더 명에서 뒤에 애니메이션은 삭제하는 것이 나아보임(상위 폴더와 중복되는 의미) + 메테리얼 폴더
-> 해결

PrototypeGun animationController에 애니메이션 이름이 모호함
-> 코드 구현상 단순히 애니메이션 한번 실행하는 거라 딱히 상관 없을거 같긴 한데, 일단 New Animation에서 Gun shooting animation으로 바꿔볼게

총 발사 애니메이션은 어디서 관리하는지?
-> 별도의 관리는 없고 그냥 애니메이션을 한번 실행하는 게 끝임,

GunHandlePrototype은 왜 2개인지? 다른 용도라면 이름 구분 필요
-> 아아 그거 그냥 프로토타입 건 꾸밀때 사용한 머터리얼 차이인데, 하나는 건의 총구 부분 머터리얼이고, 다른 하나는 나머지 몸체에 쓰인 머터리얼임, 일단 이름 구분 지음

Mirror 프리팹에 포함된 카메라의 용도는 무엇인지?
-> 그거 거울 구색은 내고 싶어서 반대편을 비추는 카메라일 뿐임, 거울에 렌더 텍스쳐가 있고, 이 렌더 텍스처랑 원하는 방향을 비추는 카메라를 연결시키면 해당 물체의 머터리얼 자체가 카메라가 비추는 모습을 보여줌.
그냥 거울 반사 기능 테스트 해보고 싶으면 레이어만 바꿔도 무방함

GrassSound 프리팹은 사운드만 가지고 있는것 같은데 다른 방식으로 관리할 수는 없는지?
-> 이거 안에 Sound Manager라는 클래스가 있는데, 나중에 원하는 Sound 관련된 기능을 담는 용도로 쓸 클래스라서 일단 냅두셨으면 좋겠습니다... (범용성있는 용도로 만들 예정)

Canvas 프리팹의 이름이 모호함
-> 수정

Assets/Resources 폴더는 빌드때 반드시 포함되며, 코드에서도 접근할 수 있는 폴더인데, 반드시 텍스쳐가 이 폴더 안에 있어야 하는지? 아니라면 따로 Assets/Textures 폴더를 만드는 것은 어떤지?
-> 나도 ㄹㅇ 만들까...? 하다가 걍 떄려박은건데 오히려 ㄳ


천재적인 피드백 감사합니다

fluetanol committed:
PR Feedback, Folder structre insertion

Taeseung 폴더 아래에 Script라는 폴더는 필요 없어보임(상위 폴더와 의미 중복)
-> 해결

Taeseung 폴더 아래에 shader 폴더는 Assets/Shader를 생성해서 옮기는 것이 좋아보임
-> 해결, 오히려 좋음

Assets/Scenes/Weapon_SampleScene/SampleScene.unity 이 씬은 사용하는 씬인건지?
-> 더미였음 삭제함

애니메이션 폴더 내의 Weapon_PrototypeAnimation 폴더 명에서 뒤에 애니메이션은 삭제하는 것이 나아보임(상위 폴더와 중복되는 의미) + 메테리얼 폴더
-> 해결

PrototypeGun animationController에 애니메이션 이름이 모호함
-> 코드 구현상 단순히 애니메이션 한번 실행하는 거라 딱히 상관 없을거 같긴 한데, 일단 New Animation에서 Gun shooting animation으로 바꿔볼게

총 발사 애니메이션은 어디서 관리하는지?
-> 별도의 관리는 없고 그냥 애니메이션을 한번 실행하는 게 끝임,

GunHandlePrototype은 왜 2개인지? 다른 용도라면 이름 구분 필요
-> 아아 그거 그냥 프로토타입 건 꾸밀때 사용한 머터리얼 차이인데, 하나는 건의 총구 부분 머터리얼이고, 다른 하나는 나머지 몸체에 쓰인 머터리얼임, 일단 이름 구분 지음

Mirror 프리팹에 포함된 카메라의 용도는 무엇인지?
-> 그거 거울 구색은 내고 싶어서 반대편을 비추는 카메라일 뿐임, 거울에 렌더 텍스쳐가 있고, 이 렌더 텍스처랑 원하는 방향을 비추는 카메라를 연결시키면 해당 물체의 머터리얼 자체가 카메라가 비추는 모습을 보여줌.
그냥 거울 반사 기능 테스트 해보고 싶으면 레이어만 바꿔도 무방함

GrassSound 프리팹은 사운드만 가지고 있는것 같은데 다른 방식으로 관리할 수는 없는지?
-> 이거 안에 Sound Manager라는 클래스가 있는데, 나중에 원하는 Sound 관련된 기능을 담는 용도로 쓸 클래스라서 일단 냅두셨으면 좋겠습니다... (범용성있는 용도로 만들 예정)

Canvas 프리팹의 이름이 모호함
-> 수정

Assets/Resources 폴더는 빌드때 반드시 포함되며, 코드에서도 접근할 수 있는 폴더인데, 반드시 텍스쳐가 이 폴더 안에 있어야 하는지? 아니라면 따로 Assets/Textures 폴더를 만드는 것은 어떤지?
-> 해결

fluetanol committed:
Code Refactoring

엉망이었던 코드 컨벤션 통일,
다만 아직 벡터관련 규칙은 어떻게 해야 할 지 모르겠어서(vs, os같은거) 그부분은 안건들고
카멜, 파스칼 표기법 등은 맞춰봤습니다.

추가로 여러분은 리팩토링후 초기화가 되어버린 인스펙터를 조심하십쇼...

fluetanol committed:
(핫픽스) 오류 수정

리팩토링 후 레이저 오브젝트가 빛이 입혀지지 않는 문제 수정

taehyeoon committed:
Style: chanage code encoding 949 to utf8(65001)

taehyeoon committed:
Fix: modify code to match code convention

packages-lock에 자동 추가됨
com.autodesk.fbx
com.unity.feature.worldbuilding
com.unity.formats.fbx
com.unity.polybrush
com.unity.probuilder
com.unity.settings-manager
com.unity.terrain-tools

fluetanol committed:
Merge pull request #5 from NewSuperLightBrothers/codeStyle

Change file encoding format & modify code convention

taehyeoon reviewed (approved):
코드 컨벤션 통일 확인

fluetanol committed:
Fix: Inspector Connection

리팩토링으로 인한 인스펙터 연결 끊김 문제 해소!
